### DISCUSSION TITLE
Speed improvements

### DISCUSSION THREAD

**Discussion by stefankinell** [karttapullautin/karttapullautin#11] [General]
I ran a few batch tests yesterday with the new cliffs-rust implementation. My 100 laz-files of Kiruna area resulted in between 79% to 88% of time spent vs last version. And compared to original perl I came down to about 49% total time. Last iteration I landed at 55%.

So good, you are still shaving off time!

I assume how much time you save depends on how much cliffs there are. When I have used KP before the time spent on executing a batch of 200 laz files much depends on what type of terrain it is.

Nice work!

**Comment by jaluzdi** [comment]
I will write on this Discussion thread to reuse it, and not to open a new one :-)

Two days ago I ran some individual test on 4 tiles (2 km x 2 Km) from Spanish LIDAR year 2012 (0.5 points/square meter)
Tile file size varied from 130MB to 173 MB
The terrain processed was a limestone karst with quite a lot of depressions.
The tiles were in *.LAS format (not in *.LAZ)

I ran the test in 3 different environments, on the same computer (as I have both operating systems installed on it):

- rusty-pullauta on Ubuntu Linux 22.04 
- rusty-pullauta on Windows 10
- pullauta on Windows 10

The improvement of rusty-pullauta versus the old pullauta is impressive. 
In Win10 the processing time goes down from [29:29 - 40:47] to [05:28 - 09:19]
And in Linux the improvement was even better [02:59 - 04:40]

<img width="865" alt="image" src="https://user-images.githubusercontent.com/130509439/231732884-ae41490e-d2f7-4907-a5fc-50e7c04bcaa0.png">

**Comment by jaluzdi** [comment]
As batch mode in **Linux** was fixed yesterday I have made some tests using it.
I do not know if these tests I made are worth it, but I would like to share them ;-)

In `pullauta.ini` I have changed the value of the parameter `processes`, asigning different values to it, from 1 to 10 
I enabled batch mode, and I did not save temp folders. Those were my parameters related to batch mode
```
batch=1
batchoutfolder=./out
lazfolder=./in
savetempfiles=0
savetempfolders=0
``` 
In every single test, I processed 24 *.laz files (tiles of 500 meters x 500 meters, with 2 points per square meter, from Spanish LIDAR 2017). Total size of *.laz files:  585 MB

The PC in which I ran the tests is an Intel Nuc (a mini PC)
The CPU is an Intel Core i5-10210U, with 32 GB RAM
[According to the specs of the CPU it has 4 Cores and it can execute 8 Threads](https://ark.intel.com/content/www/us/en/ark/products/195436/intel-core-i510210u-processor-6m-cache-up-to-4-20-ghz.html) 

In the tests, it can be seen that the time goes down when increasing the number of processes from 1 until 4
But from 4 processes onwards, there is no improvement.

Initially I thought that as the CPU is capable of executing 8 threads, I should have seen an improvement in processing time until I reached 8 in parameter `processes`. But it has not been the case. 
I do not know if this is normal.

<img width="854" alt="image" src="https://user-images.githubusercontent.com/130509439/231772876-23260d87-05c8-4155-a36b-ab98c74a5484.png">

**Reply by stefankinell** [reply]
Interesting. I have had similar observations but not that it chruns out at number of cores to a flat line like that.

I run on two different PC:S. One with an AMD Ryzen 5 4600H. 6 cores/12 threads. And one i9-12900, 16 cores, 24 threads. I do see faster execution times when I add more threads. But one thing I learned from the i9 and original kartapullautin usage is that it seems to be the temperature of the processor that sets the limit. I am not super happy with the computer with the i9 because it is not delivering what I expected in terms of multi threads. Sometimes 12 threads on the older AMD laptop is as fast as 24 threads on the i9 desktop PC.

I'll see if I have some older meassurements where I can show / visualize for you.

**Reply by rphlo** [reply]
It is possible that if a core is at 100% capacity have one or two thread on it doesnt make much differences. (rusty-)pullata is CPU intensive.

**Reply by jaluzdi** [reply]
Repeating the same tests (processing 24 *.laz files), in **Linux**, on **batch mode** varying the parameter `processes `in `pullauta.ini`, with today's source code, there is an overall improvement with the source code modifications committed during the weekend

<img width="995" alt="image" src="https://user-images.githubusercontent.com/130509439/232623214-88d9e229-e423-493e-b25f-b95fa118f094.png">

Let's see if tomorrow I can perform the same tests under Windows, to compare them

**Reply by jaluzdi** [reply]
As @rphlo kindly updated windows `pullauta.exe` in **commit #67** yesterday ;-)
I finally did the same batch test on Windows 10 (24 *.laz tiles varying number of processes in` pullauta.ini`)

In the attached graph I compare the processing times under Linux with source code from April 12th and April 17th (xyzknolls, makecliffs improvement), and under Windows 10 with source code from April 18th (xyzknolls, makecliffs improvement)

It can be seen clearly that processing times on **Linux** are much better than on **Windows** with the same source code (April 17th and April 18th are basically the same. I think the only difference is pullauta.exe for Windows)

I am performing now the same test with Jarko's pullauta (with no rust language) and tomorrow I will update the graph as it will easily take many many hours to process.

<img width="1070" alt="image" src="https://user-images.githubusercontent.com/130509439/233048607-89fe8d71-c504-46ac-92a8-f317d3e7e43e.png">

**Comment by stefankinell** [comment]
I did some more test-runs today on batch processing. On my i9-12900 (16 cores / 24 threads) I did four runs under Linux. The only difference was that number of paralell processes were changed between 10 - 16 - 20 and 24.

First observation - 24 threads was faster than 16 or 20, but just with very little. 16/20 were at the same time and 10 was slower. But when you calculate speed per process - the 10 parallell threads actually had higher performance. Which in my case I think can be explained with a) some of the cores of the i9 are efficiency cores - so when I run only 10 cores / threads the fastest only are used, or b) there are other things in the system that restricts speed, memory, temperature, read/write to disk etc.

But this test strengthen the idea of that more threads than cores does not necessary give much higher output speed.

![image](https://user-images.githubusercontent.com/59049892/232232449-5dfe2840-ef21-4ec4-ab22-ff569720a548.png)

**Comment by jaluzdi** [comment]
@rphlo , you are doing an amazing job !!!
I have just downloaded the latest version, after the commits done during the weekend (**translate xyzknolls, perl script updated, improving perfomance in makecliffs, ...**)
I have tested the new code in **Linux**, in **batch mode**, with the same 24 *.laz files I tested on April, the 12th and 13th, with **4 processes** in `pullauta.ini`

Processing time on April 12th and 13th: `13 minutes and 16 seconds`
Processing time today: `9 minutes and 2 seconds`

It is a `4 minutes and 14 seconds `performance improvement. It is nearly a 32 % improvement
Amazing !!!

P.D. I will try to do more testing this week

**Comment by stefankinell** [comment]
Just a reflection on speed. Now I am not comparing the same files so perhaps not comparing apples to apples. But last time I measured I think I had like above 5 minutes per Swedish standard .laz-file on my Mac. Now when I am batch executing in vegeonly-mode I am doing 2 minutes 28 seconds in average. And here also including some copying of files and ziping results. And with 10 parallel processes on my Mac that means in real life one law-file every 14th second. :-D The first measurement I did with perl only on my Mac took 23 minutes... 

I am using this now when reprocessing all Swedish lidar-data with new vegetation mapping. That means in total about 60 000 laz files.

If I would have run this in perl only, that would have been a lead time of 95 days.

With vegeonly in rust - 9 days.

**Reply by rphlo** [reply]
It is indeed a great improvement.
Things to factor in are:
 - The skip of contours and cliffs specifics in `vegetation only` mode.
 - The move of the `render` function to Rust.

There is still quite some work to finally reach the goal of being free of using any of the leftover Perl code, but I am satisfied with the state reached in less than a year. 86 days of computing power/cost is definitely non negligible.

Finally, I hope this will as well encourage people to take the opportunity to start generating LiDAR maps for other countries as the barrier is much now lower on the needed computing power.
